docs/1.5/extensions/heading-permalinks.md

```diff
@@ -71,6 +73,56 @@ This controls whether the anchor is added to the beginning of the `<h1>`, `<h2>`
 
 This option sets the `title` attribute on the `<a>` tag.  This defaults to `'Permalink'`.
 
+### `slug_generator`
+
+"Slugs" are the strings used within the `href`, `name`, and `id` attributes to identify a particular permalink.
+By default, this extension will generate slugs based on the contents of the heading, just like Github-Flavored Markdown does.
+
+You can change the string that is used as the "slug" by setting the `slug_generator` option to any class that implements `SlugGeneratorInterface`.
+
+For example, if you'd like each slug to be an MD5 hash, you could create a class like this:
+
+```php
+<?php
+
+use League\CommonMark\Extension\HeadingPermalink\Slug\SlugGeneratorInterface;
+
+final class HashSlugGenerator implements SlugGeneratorInterface
+{
+    public function createSlug(string $input): string
```

Comment:
Also, is there a reason this couldn't be a static method? I get the desire to keep the interface for strongly typed objects, but I think we can have our cake and eat it too.

By default, the setting in Drupal could easily be set to a string pointing to the default slug generator:

`League\CommonMark\Extension\HeadingPermalink\Slug\DefaultSlugGenerator::createSlug`

Then this means we could also expose this via the UI rather than requiring people to manually implement an alter that injects this setting manually.

All that would be needed on your end is to check if the value is a string, contains `::`, split the string on `::` and then check if the class `is_subclass_of`. I'll put the example below.

Reply:
> I'm thinking the Heading node should also be sent along for contextual reasons

I agree, but unfortunately `SlugGeneratorInterface` was released in 1.4 so we'll have to wait until 2.0 before we can modify the parameters.

In fact, we could easily make `Heading $node` the **only** parameter in this method in 2.0, since the AST changes in 2.0 ensure that all nodes which can contain text literals have a `StringContainerInterface` to easily identify that text and pull them out, and we've even got [a static utility function](https://github.com/thephpleague/commonmark/blob/fbe6deae434fc89366308e43b8349737afa73eb8/src/Node/StringContainerHelper.php#L24) that can pull them into a string for you with a simple one-liner :wink:

So unfortunately, for BC reasons, I think we'll have to keep this interface as-is for 1.5.

> Also, is there a reason this couldn't be a static method?

Dependency injection - implementors may want to reference/use other objects in their DI container without having to instantiate them manually.  Users may also want to have a single slug generator instance used to parse multiple documents to ensure there are no colliding slugs.  Ultimately it shouldn't be the responsibility of this library to instantiate the user's objects for them.

Given that Drupal is built on Symfony's DI container, what if you had your module [automatically tag any instances](https://symfony.com/doc/current/service_container/tags.html#autoconfiguring-tags) of  `SlugGeneratorInterface` and then present a list of those tagged services in a dropdown?  The module could then grab that object via the container (service location), or perhaps even [wire up that generator into the Processor during a compiler pass](https://symfony.com/doc/current/service_container/tags.html#create-a-compiler-pass).  That will keep all the instantiation logic within the container while still giving users a way to find and select their preferred generator class via the UI.

Reply:
> So unfortunately, for BC reasons, I think we'll have to keep this interface as-is for 1.5.

You could just deprecate `League\CommonMark\Extension\HeadingPermalink\Slug\SlugGeneratorInterface` and instead just create a new interface (i.e. `League\CommonMark\Slug\SlugGeneratorInterface`) that takes a single node parameter; slug generation isn't necessarily exclusive to headers.

Reply:
I'm not sure how it should be done, but simply passing the string contents of the header isn't enough to really make it useful. At the very least we'd need to know the heading level to properly store the headers.

Reply:
> At the very least we'd need to know the heading level to properly store the headers.

There are currently three ways you could do this which should work in both ^1.4 and ^2.0:

1. Iterate the `Document` yourself - any time you see a `HeadingPermalink`, grab the URL from there and the heading level from its parent `Heading` - this is what `TableOfContentsGenerator` does today.
2. Iterate the `Document` yourself looking for a `TableOfContents` element.  Note that this may not exist if using the `placeholder` option and no placeholder was provided in the Markdown.
3. Call `TableOfContentsGenerator::generate()` on the `Document` to create a new `TableOfConents` node with all the information you'd need.  Granted, this would incur duplicate processing if you're already using the ToC extension to insert a ToC, but you'd be guaranteed to have a copy created exactly as you need it.

However...

> You could just deprecate `League\CommonMark\Extension\HeadingPermalink\Slug\SlugGeneratorInterface` and instead just create a new interface (i.e.` League\CommonMark\Slug\SlugGeneratorInterface`) that takes a single node parameter; slug generation isn't necessarily exclusive to headers.

That's a fair point.  `SlugGeneratorInterface` is only used in the constructor of a `final class`, so we could remove that type hint without breaking BC, and let the class internals decide what to do based on which interface it has been given.  And accepting any `Node` wouldn't harm things :thinking:   Let me see what I can come up with :wink:

Reply:
I've added https://github.com/thephpleague/commonmark/pull/481/commits/4d8d311aa895c0826b869ff1b91411d73c3dc995 which does pretty much what you suggested :)

Mind giving this another look when you get a chance?

Reply:
LGTM

Reply:
Awesome, TYVM!